synapse/events/presence_router.py

```diff
@@ -66,14 +136,40 @@ async def get_users_for_states(
           A dictionary of user_id -> set of UserPresenceState, indicating which
           presence updates each user should receive.
         """
-        if self.custom_presence_router is not None:
-            # Ask the custom module
-            return await self.custom_presence_router.get_users_for_states(
-                state_updates=state_updates
-            )
 
-        # Don't include any extra destinations for presence updates
-        return {}
+        # Bail out early if we don't have any callbacks to run.
+        if len(self._get_users_for_states_callbacks) == 0:
+            # Don't include any extra destinations for presence updates
+            return {}
+
+        users_for_states = {}
+        # run all the callbacks for get_users_for_states and combine the results
+        for callback in self._get_users_for_states_callbacks:
+            try:
+                result = await callback(state_updates)
+            except Exception as e:
+                logger.warning("Failed to run module API callback %s: %s", callback, e)
+                continue
+
+            if not isinstance(result, Dict):
```

Comment:
I didn't realise you could do this, but it seems to work. With that said:
- https://docs.python.org/3/library/functions.html#isinstance mentions that one can use 3.10 with a union type.
- At the bottom of the documentation for typing.TypeVar, the authors note
  >In general, [isinstance()](https://docs.python.org/3/library/functions.html#isinstance) and [issubclass()](https://docs.python.org/3/library/functions.html#issubclass) should not be used with types.